src/BlockTypes/ProductImageGallery.php

```diff
@@ -28,39 +28,142 @@ protected function get_block_type_uses_context() {
 		return [ 'query', 'queryId', 'postId' ];
 	}
 
+	/**
+	 * Based on the single-product/sale-flash.php template from the core of WooCommerce.
+	 *
+	 * @param \WC_Product $product Product instance.
+	 * @param \WP_Post    $post    Post instance.
+	 */
+	private function sale_flash( $product, $post ) {
+		if ( ! $product->is_on_sale() ) {
+			return '';
+		}
+
+		ob_start();
+		echo sprintf( '<span class="onsale">%1$s</span>', esc_html__( 'Sale!', 'woo-gutenberg-products-block' ) );
+		return ob_get_clean();
```

Comment:
No need to use an output buffer here since you have control over the output. Just return it.